types/bluebird/index.d.ts

```diff
@@ -40,6 +39,33 @@ type CatchFilter<E> = ((error: E) => boolean) | (object & E);
 type Resolvable<R> = R | PromiseLike<R>;
 type IterateFunction<T, R> = (item: T, index: number, arrayLength: number) => Resolvable<R>;
 
+type PromisifyAllKeys<T> = T extends string ? T | `${T}Async` : T;
+type GetValueForKey<T, K> = K extends keyof T ? T[K] : never;
+type WithoutLast<T extends any[]> = T extends [...infer A, any] ? A : [];
+type Last<T extends any[]> = T extends [...any[], infer L] ? L : never;
+type ExtractCallbackValueType<T> = T extends (error: any, ...data: infer D) => any ? D : never;
+type ExtractAsyncMethod<T> = T extends (...args: infer A) => any
+    ? (...arg: WithoutLast<A>) => Promise<ExtractCallbackValueType<Last<Required<A>>>[0]>
+    : never;
+type ValidPromiseOrNever<T> = ExtractAsyncMethod<T> extends (...args: any[]) => Promise<never>
```

Comment:
It'd be nice if we could prevent `ExtractAsyncMethod` from returning non-never results in the first place. For instance...

This assumes that you removed the `extends any[]` constraints from `WithoutLast` and `Last`. I'm not sure if we need them. If you feel like we should keep them I'm fine with that.

We can introduce a new mapper method:
```ts
type PromiseMethod<A, R> = R extends never ? never : (...arg: WithoutLast<A>) => Promise<R>;
```

And use it:
```ts
type ExtractAsyncMethod<T> = T extends (...args: infer A) => any
    ? PromiseMethod<A, ExtractCallbackValueType<Last<A>>[0]>
    : never;
```

Afterwards we won't have to filter out invalid results using `ValidPromiseOrNever` anymore.
This ensures that we only have a single way to signal that a promisification failed: By returning `never`. Those are then centrally filtered out in `NonNeverKeys`/`NonNeverValues`.

The type names and implementation above are just an example of course.

---

By the way: I couldn't figure out why you use `Last<Required<A>>`. In my experiments `Last<A>` works just as well.

Reply:
Without `Required<T>`, functions that have an optional argument for the callback will not be present in the final result set. Tuples now store optional information:

```
type SomeObject = {
  optionalCallback: (value: string, callback?: (error: Error, output: number) => void) => void
}

Parameters<SomeObject['optionalCallback']>
/// [value: number, callback?: ((error: Error, value: string) => void) | undefined]
```